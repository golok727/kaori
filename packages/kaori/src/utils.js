export function invariant(condition, message) {
    if (!condition) {
        throw new Error(message);
    }
}
//# sourceMappingURL=utils.js.map