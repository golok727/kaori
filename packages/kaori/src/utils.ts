export function invariant(
  condition: boolean,
  message: string
): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}

export const has_own = Object.hasOwn;
export const get_own_property_descriptor = Object.getOwnPropertyDescriptor;
export const define_property = Object.defineProperty;
