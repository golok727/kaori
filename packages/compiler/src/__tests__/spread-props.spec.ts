import { describe, it, expect } from 'vitest';
import { compile } from './utils';

describe('Spread Props', () => {
  describe('HTML Elements', () => {
    it('should handle spread on HTML element', async () => {
      const input = `
function Button() {
  const props = { class: "btn", id: "button1" };
  return <button {...props}>Click me</button>;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('spread');
      expect(output).toContain('html`');
    });

    it('should handle multiple spreads on HTML element', async () => {
      const input = `
function Button() {
  const baseProps = { class: "btn" };
  const extraProps = { id: "button1" };
  return <button {...baseProps} {...extraProps}>Click me</button>;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('spread(baseProps)');
      expect(output).toContain('spread(extraProps)');
    });

    it('should handle spread mixed with regular attributes on HTML element', async () => {
      const input = `
function Button() {
  const props = { class: "btn" };
  return <button {...props} onClick={() => {}} disabled>Click me</button>;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('spread(props)');
    });

    it('should handle spread before and after attributes on HTML element', async () => {
      const input = `
function Input() {
  const props1 = { class: "input" };
  const props2 = { id: "field1" };
  return <input type="text" {...props1} placeholder="Enter text" {...props2} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('spread(props1)');
      expect(output).toContain('spread(props2)');
    });

    it('should handle self-closing HTML element with spread', async () => {
      const input = `
function Image() {
  const imgProps = { src: "/logo.png", alt: "Logo" };
  return <img {...imgProps} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('spread(imgProps)');
    });
  });

  describe('Components', () => {
    it('should optimize single spread with no other props', async () => {
      const input = `
function Wrapper() {
  const props = { value: "test" };
  return <Component {...props} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('component(Component, props)');
      expect(output).not.toContain('mergeProps');
    });

    it('should use mergeProps for single spread with other props', async () => {
      const input = `
function Wrapper() {
  const props = { value: "test" };
  return <Component {...props} other="value" />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
      expect(output).toContain('props');
      expect(output).toContain('other');
    });

    it('should use mergeProps for multiple spreads', async () => {
      const input = `
function Wrapper() {
  const props1 = { value: "test" };
  const props2 = { id: "field1" };
  return <Component {...props1} {...props2} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps(props1, props2)');
    });

    it('should use mergeProps for spreads with multiple other props', async () => {
      const input = `
function Wrapper() {
  const props = { value: "test" };
  const moreProps = { id: "field1" };
  return <Component {...props} {...moreProps} other="a" something="b" />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
      expect(output).toContain('props');
      expect(output).toContain('moreProps');
      expect(output).toContain('other');
      expect(output).toContain('something');
    });

    it('should handle spread before attributes', async () => {
      const input = `
function Wrapper() {
  const baseProps = { class: "base" };
  return <Component {...baseProps} value={getValue()} onClick={handleClick} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
      expect(output).toContain('baseProps');
    });

    it('should handle spread after attributes', async () => {
      const input = `
function Wrapper() {
  const overrides = { class: "override" };
  return <Component value={getValue()} onClick={handleClick} {...overrides} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
      expect(output).toContain('overrides');
    });

    it('should handle spread in the middle of attributes', async () => {
      const input = `
function Wrapper() {
  const middleProps = { value: "middle" };
  return <Component first="a" {...middleProps} last="z" />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
      expect(output).toContain('first');
      expect(output).toContain('middleProps');
      expect(output).toContain('last');
    });

    it('should handle complex spread scenario with reactive props', async () => {
      const input = `
function Complex() {
  const props = { value: "test" };
  return <Component {...props} onClick={() => {}} data={getData()} count={state.count} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
      expect(output).toContain('onClick: () => {}');
      expect(output).toContain('get data()');
      expect(output).toContain('get count()');
    });

    it('should handle spread with namespaced attributes', async () => {
      const input = `
function Input() {
  const props = { value: "test" };
  return <Input {...props} prop:value={getValue()} bind:checked={state.checked} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
      expect(output).toContain('["prop:value"]');
      expect(output).toContain('["bind:checked"]');
    });

    it('should handle spread with data attributes', async () => {
      const input = `
function Component() {
  const props = { value: "test" };
  return <Input {...props} data-test-id="input1" aria-label="Field" />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
      expect(output).toContain('data-test-id');
      expect(output).toContain('aria-label');
    });

    it('should not optimize single spread when component has children', async () => {
      const input = `
function Wrapper() {
  const props = { value: "test" };
  return <Component {...props}><span>Child</span></Component>;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
      expect(output).toContain('children');
    });

    it('should handle spread with children content', async () => {
      const input = `
function Wrapper() {
  const props = { value: "test" };
  return (
    <Component {...props} other="value">
      <div>Child 1</div>
      <div>Child 2</div>
    </Component>
  );
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
      expect(output).toContain('children');
    });
  });

  describe('Mixed Scenarios', () => {
    it('should handle spreads on both HTML elements and components', async () => {
      const input = `
function Mixed() {
  const buttonProps = { class: "btn" };
  const componentProps = { value: "test" };
  return (
    <div>
      <button {...buttonProps}>Click</button>
      <Component {...componentProps} />
    </div>
  );
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('spread(buttonProps)');
      expect(output).toContain('component(Component, componentProps)');
    });

    it('should handle nested components with spreads', async () => {
      const input = `
function Nested() {
  const outerProps = { class: "outer" };
  const innerProps = { value: "inner" };
  return (
    <Outer {...outerProps}>
      <Inner {...innerProps} />
    </Outer>
  );
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
    });

    it('should handle fragment with multiple spreads', async () => {
      const input = `
function Fragment() {
  const props1 = { value: "a" };
  const props2 = { value: "b" };
  return (
    <>
      <Component1 {...props1} />
      <Component2 {...props2} other="value" />
    </>
  );
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('component(Component1, props1)');
      expect(output).toContain('mergeProps(props2');
    });

    it('should handle complex real-world example', async () => {
      const input = `
function Button(props) {
  return () => {
    <>
      <Input thing={props.value} prop:value={props.value} data-thing="asd" />
      <Button {...props} onClick={() => {}} {...somethingElse}></Button>
      <button {...props}></button>
    </>
  }
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      // Button component should use mergeProps
      expect(output).toContain('mergeProps');
      // button HTML element should use spread
      expect(output).toContain('spread(props)');
    });

    it('should handle expression spreads', async () => {
      const input = `
function Dynamic() {
  const getProps = () => ({ value: "test" });
  return <Component {...getProps()} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('component(Component, getProps())');
    });

    it('should handle member expression spreads', async () => {
      const input = `
function Member() {
  const obj = { props: { value: "test" } };
  return <Component {...obj.props} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('component(Component, obj.props)');
    });

    it('should handle conditional expression spreads', async () => {
      const input = `
function Conditional() {
  const props1 = { value: "a" };
  const props2 = { value: "b" };
  const condition = true;
  return <Component {...(condition ? props1 : props2)} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('condition ? props1 : props2');
    });
  });

  describe('Edge Cases', () => {
    it('should handle empty spread object', async () => {
      const input = `
function Empty() {
  return <Component {...{}} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
    });

    it('should handle spread with inline object', async () => {
      const input = `
function Inline() {
  return <Component {...{ value: "test", other: 123 }} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
    });

    it('should handle multiple consecutive spreads', async () => {
      const input = `
function Multiple() {
  return <Component {...props1} {...props2} {...props3} {...props4} />;
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps(props1, props2, props3, props4)');
    });

    it('should handle spread with all attribute types', async () => {
      const input = `
function AllTypes() {
  const props = { base: "value" };
  return (
    <Component
      {...props}
      onClick={handleClick}
      prop:value={getValue()}
      bool:disabled={isDisabled()}
      data-test="test"
      aria-label="Label"
      ref={ref}
    />
  );
}`;

      const output = await compile(input);
      expect(output).toMatchSnapshot();
      expect(output).toContain('mergeProps');
    });
  });
});
