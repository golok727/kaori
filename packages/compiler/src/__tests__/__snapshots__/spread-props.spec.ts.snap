// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spread Props > Components > should handle complex spread scenario with reactive props 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Complex() {
  const props = {
    value: "test"
  };
  return html\`\${component(Component, mergeProps(props, {
    onClick: () => {},
    get data() {
      return getData();
    },
    get count() {
      return state.count;
    }
  }))}\`;
}"
`;

exports[`Spread Props > Components > should handle spread after attributes 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Wrapper() {
  const overrides = {
    class: "override"
  };
  return html\`\${component(Component, mergeProps({
    get value() {
      return getValue();
    },
    onClick: handleClick
  }, overrides))}\`;
}"
`;

exports[`Spread Props > Components > should handle spread before attributes 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Wrapper() {
  const baseProps = {
    class: "base"
  };
  return html\`\${component(Component, mergeProps(baseProps, {
    get value() {
      return getValue();
    },
    onClick: handleClick
  }))}\`;
}"
`;

exports[`Spread Props > Components > should handle spread in the middle of attributes 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Wrapper() {
  const middleProps = {
    value: "middle"
  };
  return html\`\${component(Component, mergeProps({
    first: "a"
  }, middleProps, {
    last: "z"
  }))}\`;
}"
`;

exports[`Spread Props > Components > should handle spread with children content 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Wrapper() {
  const props = {
    value: "test"
  };
  return html\`\${component(Component, mergeProps(props, {
    other: "value"
  }, {
    children: [html\`<div>Child 1</div>\`, html\`<div>Child 2</div>\`]
  }))}\`;
}"
`;

exports[`Spread Props > Components > should handle spread with data attributes 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Component() {
  const props = {
    value: "test"
  };
  return html\`\${component(Input, mergeProps(props, {
    "data-test-id": "input1",
    "aria-label": "Field"
  }))}\`;
}"
`;

exports[`Spread Props > Components > should handle spread with namespaced attributes 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Input() {
  const props = {
    value: "test"
  };
  return html\`\${component(Input, mergeProps(props, {
    get ["prop:value"]() {
      return getValue();
    },
    get ["bind:checked"]() {
      return state.checked;
    }
  }))}\`;
}"
`;

exports[`Spread Props > Components > should not optimize single spread when component has children 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Wrapper() {
  const props = {
    value: "test"
  };
  return html\`\${component(Component, mergeProps(props, {
    children: html\`<span>Child</span>\`
  }))}\`;
}"
`;

exports[`Spread Props > Components > should optimize single spread with no other props 1`] = `
"import { component, html } from "kaori.js";
function Wrapper() {
  const props = {
    value: "test"
  };
  return html\`\${component(Component, props)}\`;
}"
`;

exports[`Spread Props > Components > should use mergeProps for multiple spreads 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Wrapper() {
  const props1 = {
    value: "test"
  };
  const props2 = {
    id: "field1"
  };
  return html\`\${component(Component, mergeProps(props1, props2))}\`;
}"
`;

exports[`Spread Props > Components > should use mergeProps for single spread with other props 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Wrapper() {
  const props = {
    value: "test"
  };
  return html\`\${component(Component, mergeProps(props, {
    other: "value"
  }))}\`;
}"
`;

exports[`Spread Props > Components > should use mergeProps for spreads with multiple other props 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Wrapper() {
  const props = {
    value: "test"
  };
  const moreProps = {
    id: "field1"
  };
  return html\`\${component(Component, mergeProps(props, moreProps, {
    other: "a",
    something: "b"
  }))}\`;
}"
`;

exports[`Spread Props > Edge Cases > should handle empty spread object 1`] = `
"import { component, html } from "kaori.js";
function Empty() {
  return html\`\${component(Component, {})}\`;
}"
`;

exports[`Spread Props > Edge Cases > should handle multiple consecutive spreads 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Multiple() {
  return html\`\${component(Component, mergeProps(props1, props2, props3, props4))}\`;
}"
`;

exports[`Spread Props > Edge Cases > should handle spread with all attribute types 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function AllTypes() {
  const props = {
    base: "value"
  };
  return html\`\${component(Component, mergeProps(props, {
    onClick: handleClick,
    get ["prop:value"]() {
      return getValue();
    },
    get ["bool:disabled"]() {
      return isDisabled();
    },
    "data-test": "test",
    "aria-label": "Label",
    ref: ref
  }))}\`;
}"
`;

exports[`Spread Props > Edge Cases > should handle spread with inline object 1`] = `
"import { component, html } from "kaori.js";
function Inline() {
  return html\`\${component(Component, {
    value: "test",
    other: 123
  })}\`;
}"
`;

exports[`Spread Props > HTML Elements > should handle multiple spreads on HTML element 1`] = `
"import { html, spread } from "kaori.js";
function Button() {
  const baseProps = {
    class: "btn"
  };
  const extraProps = {
    id: "button1"
  };
  return html\`<button \${spread(baseProps)} \${spread(extraProps)}>Click me</button>\`;
}"
`;

exports[`Spread Props > HTML Elements > should handle self-closing HTML element with spread 1`] = `
"import { html, spread } from "kaori.js";
function Image() {
  const imgProps = {
    src: "/logo.png",
    alt: "Logo"
  };
  return html\`<img \${spread(imgProps)} />\`;
}"
`;

exports[`Spread Props > HTML Elements > should handle spread before and after attributes on HTML element 1`] = `
"import { html, spread } from "kaori.js";
function Input() {
  const props1 = {
    class: "input"
  };
  const props2 = {
    id: "field1"
  };
  return html\`<input type="text" \${spread(props1)} placeholder="Enter text" \${spread(props2)} />\`;
}"
`;

exports[`Spread Props > HTML Elements > should handle spread mixed with regular attributes on HTML element 1`] = `
"import { html, spread } from "kaori.js";
function Button() {
  const props = {
    class: "btn"
  };
  return html\`<button \${spread(props)} @click=\${() => {}} disabled=\${true}>Click me</button>\`;
}"
`;

exports[`Spread Props > HTML Elements > should handle spread on HTML element 1`] = `
"import { html, spread } from "kaori.js";
function Button() {
  const props = {
    class: "btn",
    id: "button1"
  };
  return html\`<button \${spread(props)}>Click me</button>\`;
}"
`;

exports[`Spread Props > Mixed Scenarios > should handle complex real-world example 1`] = `
"import { component, html, spread, mergeProps } from "kaori.js";
function Button(props) {
  return () => {
    [html\`\${component(Input, {
      get thing() {
        return props.value;
      },
      get ["prop:value"]() {
        return props.value;
      },
      "data-thing": "asd"
    })}\`, html\`\${component(Button, mergeProps(props, {
      onClick: () => {}
    }, somethingElse))}\`, html\`<button \${spread(props)}></button>\`];
  };
}"
`;

exports[`Spread Props > Mixed Scenarios > should handle conditional expression spreads 1`] = `
"import { component, html } from "kaori.js";
function Conditional() {
  const props1 = {
    value: "a"
  };
  const props2 = {
    value: "b"
  };
  const condition = true;
  return html\`\${component(Component, condition ? props1 : props2)}\`;
}"
`;

exports[`Spread Props > Mixed Scenarios > should handle expression spreads 1`] = `
"import { component, html } from "kaori.js";
function Dynamic() {
  const getProps = () => ({
    value: "test"
  });
  return html\`\${component(Component, getProps())}\`;
}"
`;

exports[`Spread Props > Mixed Scenarios > should handle fragment with multiple spreads 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Fragment() {
  const props1 = {
    value: "a"
  };
  const props2 = {
    value: "b"
  };
  return [html\`\${component(Component1, props1)}\`, html\`\${component(Component2, mergeProps(props2, {
    other: "value"
  }))}\`];
}"
`;

exports[`Spread Props > Mixed Scenarios > should handle member expression spreads 1`] = `
"import { component, html } from "kaori.js";
function Member() {
  const obj = {
    props: {
      value: "test"
    }
  };
  return html\`\${component(Component, obj.props)}\`;
}"
`;

exports[`Spread Props > Mixed Scenarios > should handle nested components with spreads 1`] = `
"import { component, html, mergeProps } from "kaori.js";
function Nested() {
  const outerProps = {
    class: "outer"
  };
  const innerProps = {
    value: "inner"
  };
  return html\`\${component(Outer, mergeProps(outerProps, {
    get children() {
      return html\`\${component(Inner, innerProps)}\`;
    }
  }))}\`;
}"
`;

exports[`Spread Props > Mixed Scenarios > should handle spreads on both HTML elements and components 1`] = `
"import { component, html, spread } from "kaori.js";
function Mixed() {
  const buttonProps = {
    class: "btn"
  };
  const componentProps = {
    value: "test"
  };
  return html\`<div><button \${spread(buttonProps)}>Click</button>\${component(Component, componentProps)}</div>\`;
}"
`;
